### PR TITLE
Add endpoint to log client events

### PR DESCRIPTION
**Pull request by crobibero**
Fixes https://github.com/jellyfin/jellyfin/issues/5646

This logs all events to files separated by the `ClientName` parameter as the following to the file `log_{ClientName}_{yyyymmdd}.log`

```
[2021-04-26 12:21:00.108 +00:00] [Trace] [Jellyfin-Sample-Client:1.0.0]: UserId: 0000000-0000-0000-0000-000000000000 DeviceId:  8eec4223-9280-454f-a4ee-bc978c1007e4 
This is the log message 
```

Updated openapi: [openapi_client_event.json.txt](https://github.com/jellyfin/jellyfin/files/6377234/openapi_client_event.json.txt)

Note- these endpoints don't currently require authentication, so that may change depending on feedback.

### CODE REVIEW

**Comment by dkanada** [comment]
Can we put the client name first so they're all in a single location on the filesystem? Our ffmpeg logs have a similar pattern, which I find quite convenient.

**Review by BaronGreenback — APPROVED** [review]

**Review by Ullmie02 — COMMENTED** [review]
Shouldn't the api endpoint require authorization?

**Review by Ullmie02 — APPROVED** [review]

**Review by nielsvanvelzen — APPROVED** [review]

**Comment by nielsvanvelzen** [comment]
@crobibero can you rebase to fix the merge conflict and so we can see if #6308 works?

**Comment by crobibero** [comment]
@nielsvanvelzen 

>  Error: Resource not accessible by integration
Error: See this action's readme for details about this error

File: Jellyfin.Api/Controllers/ClientLogController.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        /// <param name="clientLogEventDto">The client log dto.</param>
+        /// <response code="204">Event logged.</response>
+        /// <response code="403">Event logging disabled.</response>
+        /// <returns>Submission status.</returns>
+        [HttpPost]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        public async Task<ActionResult> LogEvent([FromBody] ClientLogEventDto clientLogEventDto)
+        {
+            if (!_serverConfigurationManager.Configuration.AllowClientLogUpload)
```
Comment (cvium) [inline Jellyfin.Api/Controllers/ClientLogController.cs:55]: Could also have been a policy, but I'm also not a fan of the policies.

File: Jellyfin.Api/Controllers/ClientLogController.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                return Forbid();
+            }
+
+            var authorizationInfo = await _authorizationContext.GetAuthorizationInfo(Request)
```
Comment (cvium) [inline Jellyfin.Api/Controllers/ClientLogController.cs:60]: Technically, the information you want is already present in `HttpContext.User.Claims`

Comment (cvium) [reply Jellyfin.Api/Controllers/ClientLogController.cs:61]: Is this resolved though?

**Comment by nielsvanvelzen** [comment]
> @nielsvanvelzen
> 
> > Error: Resource not accessible by integration
> > Error: See this action's readme for details about this error
> 
> > Note: In public repositories this action does not work in pull_request workflows when triggered by forks. Any attempt will be met with the error, Resource not accessible by integration. This is due to token restrictions put in place by GitHub Actions. Private repositories can be configured to enable workflows from forks to run without restriction. See here for further explanation. Alternatively, use the pull_request_target event to comment on pull requests.

This should be fixed now, do you want to rebase again so we can test it?

**Comment by crobibero** [comment]
@jellyfin-bot rebase

File: MediaBrowser.Controller/ClientEvent/ClientEventLogger.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        /// <inheritdoc />
+        public async Task<string> WriteDocumentAsync(string clientName, string clientVersion, Stream fileContents)
+        {
+            var fileName = $"upload_{clientName}_{clientVersion}_{DateTime.UtcNow:yyyyMMddHHmmss}.log";
+            var logFilePath = Path.Combine(_applicationPaths.LogDirectoryPath, fileName);
+            await using var fileStream = new FileStream(logFilePath, FileMode.CreateNew, FileAccess.Write, FileShare.None);
+            await fileContents.CopyToAsync(fileStream).ConfigureAwait(false);
+            await fileStream.FlushAsync().ConfigureAwait(false);
```
Comment (cvium) [inline MediaBrowser.Controller/ClientEvent/ClientEventLogger.cs:53]: `DisposeAsync` should already be flushing

File: MediaBrowser.Controller/ClientEvent/ClientEventLogger.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                clientLogEvent.UserId ?? Guid.Empty,
+                clientLogEvent.DeviceId,
+                Environment.NewLine,
+                clientLogEvent.Message);
+        }
+
+        /// <inheritdoc />
+        public async Task<string> WriteDocumentAsync(string clientName, string clientVersion, Stream fileContents)
+        {
+            var fileName = $"upload_{clientName}_{clientVersion}_{DateTime.UtcNow:yyyyMMddHHmmss}.log";
```
Comment (cvium) [inline MediaBrowser.Controller/ClientEvent/ClientEventLogger.cs:48]: What happens if there's a hiccup and it sends multiple at once? Or if two separate clients with the same name send logs at the same time?

Comment (crobibero) [reply MediaBrowser.Controller/ClientEvent/ClientEventLogger.cs]: Great catch, that wouldn't be very good- updated

**Review by cvium — APPROVED** [review]

**Pull request merged**